# Domain-based gene mining for phylogenetic database construction

This repository contains one Python script: filo.py

## Phylogenetic analysis toolkit
The filo.py script contains two functions:

#### pfam
This function receives a list of fasta files and uses the hmmscan program to generate a database for each input file of the domains present in each sequence contained in the files.

To call this function:

```python
from filo import pfam

seq_list = ['fasta1', 'fasta2', 'fasta3']
seq_path = '/path/to/directory_of_fasta'
dicio = {'specie': ['code', 'abreviation']}
pfam_path = '/path/to/pfam_database'
regex = 'regular_expression_to_identify_files'

print('\n Retriving database')
pfam(seq_list, seq_path, dicio, pfam_path, regex)
``` 

#### analysis
This function receives the database generated by the pfam function and filters the sequences that contain the domain specified by the user. Then, the filtered sequences are analyzed by the pepstats program from the emboss package, deeploc2, mafft, CiAlign and iqtree2. At the end, a database is generated containing the sequence data, such as species, genome assembly, sequence ID, sequence length, isoelectric point, subcellular localization, and also a phylogenetic tree separated in another file.

To call this function:

```python
from filo import analysis

outdir = 'path/to/output_directory'
domain = ['domain1', 'domain2']
seq_list = ['fasta1', 'fasta2', 'fasta3']
seq_path = '/path/to/directory_of_fasta'
dicio = {'specie': ['code', 'abreviation']}
regex = 'regular_expression_to_identify_files'

print('\n Starting gene prospecting analyses')
analysis(outdir, domain, seq_list, seq_path, dicio, regex)
``` 

## Requirements
To use the filo.py script, you need to have the following programs installed:

- HMMER (hmmscan)
- EMBOSS (pepstats)
- deeploc2
- Signalp6
- DeepTMHMM
- MAFFT
- CiAlign
- IQ-TREE

You also need to have Python 3.x installed, along with the following Python modules:

- biolib
- biopython
- pandas

## Attention

To use the filo.py script, make sure to have the necessary programs installed.

Note that this script was developed for research purposes and may require modifications to adapt to different scenarios.
